nursery/refund.go

```diff
@@ -0,0 +1,130 @@
+package nursery
+
+import (
+	"fmt"
+	"github.com/BoltzExchange/boltz-client/boltz"
+	"github.com/BoltzExchange/boltz-client/boltzrpc"
+	"github.com/BoltzExchange/boltz-client/database"
+	"github.com/BoltzExchange/boltz-client/logger"
+	"strconv"
+)
+
+func (nursery *Nursery) startBlockListener(currency boltz.Currency) {
+	blockNotifier := nursery.registerBlockListener(currency)
+
+	go func() {
+		for newBlock := range blockNotifier {
+			if nursery.stopped {
+				return
+			}
+			swaps, err := nursery.database.QueryRefundableSwapsForBlockHeight(newBlock.Height, currency)
+			if err != nil {
+				logger.Error("Could not query refundable Swaps: " + err.Error())
+				continue
+			}
+
+			chainSwaps, err := nursery.database.QueryRefundableChainSwaps(newBlock.Height, currency)
+			if err != nil {
+				logger.Error("Could not query refundable Swaps: " + err.Error())
+				continue
+			}
+
+			if len(swaps) > 0 || len(chainSwaps) > 0 {
+				logger.Info("Found " + strconv.Itoa(len(swaps)) + " Swaps to refund at height " + strconv.FormatUint(uint64(newBlock.Height), 10))
+
+				if err := nursery.RefundSwaps(swaps, chainSwaps, false); err != nil {
+					logger.Error("Could not refund Swaps: " + err.Error())
+				}
+			}
+		}
+	}()
+}
+
+func (nursery *Nursery) RefundSwaps(swaps []database.Swap, chainSwaps []database.ChainSwap, cooperative bool) error {
+	var currency boltz.Currency
+	var refundedSwaps []database.Swap
+	var refundedChainSwaps []database.ChainSwap
+	var outputs []*Output
+
+	for _, swap := range swaps {
+		currency = swap.Pair.From
+		refundOutput, err := nursery.getRefundOutput(&swap)
+		if err != nil {
+			return fmt.Errorf("could not get refund output of swap %s: %v", swap.Id, err)
+		}
+
+		refundOutput.Cooperative = cooperative
+		refundedSwaps = append(refundedSwaps, swap)
+		outputs = append(outputs, refundOutput)
+	}
+	for _, chainSwap := range chainSwaps {
+		currency = chainSwap.Pair.From
+		refundOutput, err := nursery.getChainSwapRefundOutput(&chainSwap)
+		if err != nil {
+			return fmt.Errorf("could not get refund output for chain swap %s: %v", chainSwap.Id, err)
+		}
+
+		refundOutput.Cooperative = cooperative
+		refundedChainSwaps = append(refundedChainSwaps, chainSwap)
+		outputs = append(outputs, refundOutput)
+	}
```

Comment:
Any chance we could make them adhere to a common interface to avoid the code duplication?